rotkehlchen/chain/ethereum/modules/balancer/decoder.py

```diff
@@ -249,14 +249,134 @@ def _maybe_enrich_balancer_v1_events(
+        """
+        deposited_assets = set()
+
+        for event in decoded_events:
+            if (
+                event.counterparty == CPT_BALANCER_V1 and
+                event.event_type == HistoryEventType.DEPOSIT and
+                event.event_subtype == HistoryEventSubType.DEPOSIT_ASSET
+            ):
+                deposited_assets.add(event.asset)
+            elif (
+                event.counterparty == CPT_BALANCER_V1 and
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.RECEIVE_WRAPPED and
+                event.asset in deposited_assets
+            ):
+                # in this case we got refunded one of the assets deposited
+                event.event_subtype = HistoryEventSubType.REMOVE_ASSET
+                asset = event.asset.resolve_to_asset_with_symbol()
+                event.notes = f'Refunded {event.balance.amount} {asset.symbol} after depositing in Balancer V1 pool'  # noqa: E501
+
+        return decoded_events
+
+    def _check_deposits_withdrawals_v1(
+            self,
+            transaction: EvmTransaction,  # pylint: disable=unused-argument
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+    ) -> list[HistoryBaseEntry]:
+        """
+        Check for accounting in v1 that the deposits/withdrawals events have the needed information
+        to process them during accounting.
+        """
+        receive_event, return_event = None, None
+        withdrawal_events, deposit_events = [], []
+        balancer_v1_event = False
+        for event in decoded_events:
+            if event.counterparty != CPT_BALANCER_V1:
+                continue
+
+            balancer_v1_event = True
+            if (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.RECEIVE_WRAPPED
+            ):
+                receive_event = event
+            elif (
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.RETURN_WRAPPED
+            ):
+                return_event = event
+            elif (
+                (
+                    event.event_type == HistoryEventType.DEPOSIT and
+                    event.event_subtype == HistoryEventSubType.DEPOSIT_ASSET
+                ) or
+                (
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.event_subtype == HistoryEventSubType.REMOVE_ASSET
+                )
+            ):
+                deposit_events.append(event)
+            elif (
+                event.event_type == HistoryEventType.WITHDRAWAL and
+                event.event_subtype == HistoryEventSubType.REMOVE_ASSET
+            ):
+                withdrawal_events.append(event)
+
+        if balancer_v1_event is False:
```

Comment:
What happens if there is multiple receive, return or withdrawal events in a transaction? Your approach will have the last one only at this point.

Reply:
Not sure if there is a multiple deposit case in the wild but then this is a flaw that we also have in the curve decoder.

Reply:
I have handled this case now

Reply:
have you handled this in both cases?